File: LeetCode_Algorithm/135(Candy)/135.py
class Solution:
    def candy(self, ratings: list[int]) -> int:
        # sol1 172ms 62% 19.5MB 19%
        n = len(ratings)
        left = [1] * n; right = [1] * n
        for l in range(1, n):
            if ratings[l] > ratings[l-1]:
                left[l] = left[l-1] + 1
        for r in range(n-2, -1, -1):
            if ratings[r] > ratings[r+1]:
                right[r] = right[r+1] + 1
        ans = 0
        for i in range(n):
            ans += max(left[i], right[i])
        return ans

print(Solution().candy([1,0,2]))
print(Solution().candy([1,2,2]))
print(Solution().candy([1,2,3]))
print(Solution().candy([3,2,1]))
print(Solution().candy([0,4,4,5,2,1]))




            

File: LeetCode_Algorithm/135(Candy)/135.js
var candy = function(ratings) {
  // sol1 하나밖에 없음 78ms 41% 46.4MB 23%
  let n = ratings.length
  let left = Array(n).fill(1)
  let right = Array(n).fill(1)
  for (let i = 1; i < n; i++) {
    if (ratings[i] > ratings[i-1]) {
      left[i] = left[i-1] + 1
    }
  }
  for (let i = n-2; i >= 0; i--) {
    if (ratings[i] > ratings[i+1]) {
      right[i] = right[i+1] + 1
    }
  }
  let ans = 0
  for (let i = 0; i < n; i++) {
    ans += Math.max(left[i], right[i])
  }
  return ans
};
